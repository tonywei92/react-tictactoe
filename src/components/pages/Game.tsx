/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable react-hooks/exhaustive-deps */
/** @jsx jsx */
import { jsx } from '@emotion/core';
import tw, { css } from 'twin.macro';
import React, { useEffect } from 'react';
import MainLayout from 'components/layouts/MainLayout';
import TicTacToe from 'components/TicTacToe';
import Modal from 'components/UI/Modal';
import Button from 'components/UI/Button';
import NavBar from 'components/UI/NavBar';
import ProgressBar from 'components/UI/ProgressBar';
import ScoreBoard from 'components/UI/ScoreBoard';
import Smile from 'components/UI/icons/Smile';
import Suprised from 'components/UI/icons/Surprised';
import useTimer from 'components/hooks/useTimer';
import { GameStoreType } from 'stores/GameStore';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { XHolder, OHolder } from 'components/TicTacToe/Pawn';

const { useTicTacToe } = TicTacToe;
const bellSound = require('../../assets/bell.mp3');
const failSound = require('../../assets/fail.mp3');

const animationPing = css`
  ${tw`absolute`}
  animation: ping1 0.8s cubic-bezier(0.4, 0, 0.6, 1) infinite;
  left: -6px;
  top: -30px;
  @keyframes ping1 {
    0% {
      transform: scale(1);
      opacity: 1;
    }
    100% {
      transform: scale(1.6);
      opacity: 0;
    }
  }
`;

const Game = () => {
  const gridSize = useSelector((state: GameStoreType) => state.gridSize);
  const winningConditionSize = useSelector(
    (state: GameStoreType) => state.winningSize
  );
  const timeLimit = useSelector((state: GameStoreType) => state.timeLimit);
  const history = useHistory();
  const ticTacToe = useTicTacToe(gridSize, winningConditionSize);
  const {
    winner,
    retryGame,
    playerOScore,
    playerXScore,
    tie,
    nextTurn,
  } = ticTacToe;

  const timer = useTimer(timeLimit, () => {
    new Audio(failSound).play();
    timer.resetTimer();
    timer.startTimer();
    nextTurn();
  });

  const restartGame = (): void => {
    retryGame();
    timer.resetTimer();
    timer.startTimer();
  };

  const handleEmptyHolderClick = (): void => {
    timer.resetTimer();
    timer.startTimer();
  };

  // start timer on game start
  useEffect(() => {
    timer.startTimer();
  }, []);

  // start timer on game over
  useEffect(() => {
    if (winner.winner) {
      timer.resetTimer();
      new Audio(bellSound).play();
    }
  }, [winner.winner]);

  const modalButtons = [
    <Button
      size="small"
      key="retry"
      type="button"
      onClick={(): void => {
        restartGame();
      }}
    >
      Retry
    </Button>,
    <Button
      tw="ml-4"
      size="small"
      key="home"
      type="button"
      danger
      onClick={(): void => {
        history.push('/');
      }}
    >
      Back to home
    </Button>,
  ];

  const TicTacToeGameRender = React.useMemo(
    () => (
      <TicTacToe
        tictactoe={ticTacToe}
        onEmptyHolderClick={handleEmptyHolderClick}
      />
    ),
    [ticTacToe, handleEmptyHolderClick]
  );

  return (
    <MainLayout>
      <Modal
        visible={!!winner.winner}
        buttons={modalButtons}
        icon={winner.winner === 'tie' ? <Suprised /> : <Smile />}
      >
        {winner.winner === 'tie' ? (
          'Uh oh, the game is tie!'
        ) : (
          <div tw="flex items-center">
            <div>
              Congratulations, the winner is&nbsp;
              <div tw="w-12 h-8 inline-block align-middle">
                {winner.winner === 'x' ? <XHolder /> : <OHolder />}
              </div>
            </div>
          </div>
        )}
      </Modal>
      <NavBar
        leftButton={
          <Button size="small" danger onClick={(): void => history.push('/')}>
            Home
          </Button>
        }
        rightButton={
          <Button size="small" onClick={(): void => restartGame()}>
            Restart
          </Button>
        }
      />

      <ScoreBoard
        scores={[
          { label: 'Scored by X', score: playerXScore },
          { label: 'Tie', score: tie },
          { label: 'Scored by O', score: playerOScore },
        ]}
      />

      <div tw="w-full md:w-8/12 lg:w-6/12 p-4 flex flex-col justify-center items-center">
        <div css={tw`flex items-center justify-center mb-4 relative w-full`}>
          {!winner.winner ? (
            <div tw="w-16 mr-4 h-12 absolute" css={animationPing}>
              Hurry&nbsp;up!
              {ticTacToe.getPlayerTurn() === 'x' ? <XHolder /> : <OHolder />}
            </div>
          ) : null}
          <div tw="w-16 mr-4 h-12">
            {ticTacToe.getPlayerTurn() === 'x' ? <XHolder /> : <OHolder />}
          </div>
          <div tw="w-full h-full flex items-stretch justify-center items-center">
            <ProgressBar progress={timer.percentProgress} />
          </div>
        </div>
      </div>
      {TicTacToeGameRender}
    </MainLayout>
  );
};

export default Game;
