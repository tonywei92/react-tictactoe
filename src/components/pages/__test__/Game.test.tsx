/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import gameStore from 'stores/GameStore';
import Game from '../Game';

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <Provider store={gameStore}>
      <Game />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
