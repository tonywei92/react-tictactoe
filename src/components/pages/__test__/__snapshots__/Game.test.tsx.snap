// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<div
  className="css-y7ld5l-MainLayout"
>
  <div
    className="css-5g8jmz-Modal"
    style={
      Object {
        "top": "-200px",
      }
    }
  >
    <div
      className="css-qfiafw-Modal"
    >
      <div
        className="css-pmzcs3-Modal"
      >
        <svg
          className="icon icon-tabler icon-tabler-mood-smile"
          fill="none"
          height="44"
          stroke="#2c3e50"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="1.5"
          viewBox="0 0 24 24"
          width="44"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            stroke="none"
          />
          <circle
            cx="12"
            cy="12"
            r="9"
          />
          <line
            x1="9"
            x2="9.01"
            y1="10"
            y2="10"
          />
          <line
            x1="15"
            x2="15.01"
            y1="10"
            y2="10"
          />
          <path
            d="M9.5 15a3.5 3.5 0 0 0 5 0"
          />
        </svg>
      </div>
      <div>
        <div
          className="css-1ji4y71-Game"
        >
          <div>
            Congratulations, the winner is 
            <div
              className="css-16nfzq1-Game"
            >
              <div
                className="css-lcxxtz-OHolder"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr />
    <div
      className="css-1tkpjih-Modal"
    >
      <button
        className="css-rvw9k7"
        onClick={[Function]}
        type="button"
      >
        Retry
      </button>
      <button
        className="css-1lmys2w-Game"
        onClick={[Function]}
        type="button"
      >
        Back to home
      </button>
    </div>
  </div>
  <nav
    className="css-pt2wga-NavBar"
  >
    <div
      className="css-ij8jdi-NavBar"
    >
      <div
        className="css-1x6gbcb-NavBar"
      >
        <button
          className="css-19j81ss"
          onClick={[Function]}
          type="button"
        >
          Home
        </button>
      </div>
      <div
        className="css-200zck-GameLogo"
      >
        TicTacToe v
      </div>
    </div>
  </nav>
  <div
    className="css-7qv59r-ScoreBoard"
  >
    <div
      className="css-1wuj7j"
    >
      <div
        className="css-gcsfdf"
      >
        0
      </div>
      <div
        className="css-zx2i4y"
      >
        Scored by X
      </div>
    </div>
    <div
      className="css-1wuj7j"
    >
      <div
        className="css-gcsfdf"
      >
        0
      </div>
      <div
        className="css-zx2i4y"
      >
        Tie
      </div>
    </div>
    <div
      className="css-1wuj7j"
    >
      <div
        className="css-gcsfdf"
      >
        0
      </div>
      <div
        className="css-zx2i4y"
      >
        Scored by O
      </div>
    </div>
  </div>
  <div
    className="css-17fo55a-Game"
  >
    <div
      className="css-1vm4eyw-Game"
    >
      <div
        className="css-yij87g-Game"
      >
        Hurry up!
        <div
          className="css-19qisku-XHolder"
        />
      </div>
      <div
        className="css-12dfc0p-Game"
      >
        <div
          className="css-19qisku-XHolder"
        />
      </div>
      <div
        className="css-gipbfv-Game"
      >
        <div
          className="css-s27jqe-ProgressBar"
        >
          <div
            className="css-1gt18hc-ProgressBar"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-1va1icm-Board"
  >
    <div
      className="gameContainerx css-m82zkr-Board"
    >
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      >
        <div
          aria-label="grid"
          aria-readonly={true}
          className="ReactVirtualized__Grid css-1jj07fx"
          onScroll={[Function]}
          role="grid"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "ltr",
              "height": 0,
              "overflowX": "hidden",
              "overflowY": "hidden",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
  <div
    className="css-158bw6j-MainLayout"
  >
    <span>
      Made with 
      <span
        aria-label="love symbol"
        role="img"
      >
        ❤️
      </span>
        by Tony Song
    </span>
  </div>
</div>
`;
